### PR TITLE
(PF-2437) Allow r10k to authenticate to the Forge

### PR DESCRIPTION
**Pull request by binford2k**
In the near future, the Forge will be adding support for authenticated
module downloads. This extends support for token based authentication to
that use case.

### CODE REVIEW

**Review by Magisus — APPROVED** [review]

**Comment by Magisus** [comment]
Oh cool, good timing on us adding this support then.

File: doc/dynamic-environments/configuration.mkd
Comment (mwaggett) [inline doc/dynamic-environments/configuration.mkd:171]:
```suggestion
You will need to prepend your token with 'Bearer ' to authenticate to the Forge or when using Artifactory as your Forge server.
```

File: spec/unit/action/runner_spec.rb
Comment (mwaggett) [inline spec/unit/action/runner_spec.rb:314]: Would it be valuable to test that setting `authorization_token` now _does_ work if `baseurl` isn't set?

**Review by mwaggett — COMMENTED** [review]
Have you done any manual testing?

Is there Forge work that needs to land before this can be released?

File: spec/unit/action/runner_spec.rb
Comment (Magisus) [reply spec/unit/action/runner_spec.rb:314]: Considering how mocked up this all is, not really IMO.

**Comment by Magisus** [comment]
Is it even possible to do manual testing right now, or does the forge not yet support token auth?

**Comment by binford2k** [comment]
There is Forge WIP that will need to land before this is actually functional. It shouldn't actually break anything, so it won't _hurt_ anything to release early, but I'd wait until @carabasdaniel gives the go-ahead.

**Comment by mwaggett** [comment]
okay cool. yeah @binford2k I'd like to wait for that to land so that this can be manually tested before merging.

**Comment by caseywilliams** [comment]
This won't be ready for testing until PF-2447 is done - I've updated the jira ticket for this PR to reflect that. I'd prefer to wait until we can test with real data before we merge these changes.

**Comment by geoffnichols** [comment]
@binford2k and @mwaggett - when y'all get a chance to take a look at this again, it looks like PF-2447 was resolved yesterday!

**Comment by mwaggett** [comment]
awesome, @binford2k would you mind rebasing and then doing some manual testing?

File: doc/dynamic-environments/configuration.mkd
Comment (caseywilliams) [reply doc/dynamic-environments/configuration.mkd:171]: I suggest waiting to merge this until https://github.com/puppetlabs/forge-ruby/pull/89 is merged and released, since it obviates the need to include 'Bearer'

Comment (Magisus) [reply doc/dynamic-environments/configuration.mkd:171]: Looks like that should work for Artifactory-based Forge too?

Comment (caseywilliams) [reply doc/dynamic-environments/configuration.mkd:171]: I think so, although I haven't checked that manually - I did manually verify just now that the version of forge-ruby I just released (3.1.0) will allow this to work with the forge API without prepending `Bearer ` in the r10k.yaml setting. Otherwise, the only thing stopping r10k from authenticating before was the check for a `baseurl` that's removed in this PR.

Unless I'm missing something, it seems like the options here are to:
* Update this README to suggest specifying `Bearer <your-key>` as the `authorization_token` setting for now (regardless of whether you're authenticating to the forge or to artifactory), and leave the forge-ruby dependency as-is
or
* Update the forge-ruby dependency to 3.1.0 and don't mention `Bearer` at all - the  biggest impact from this would probably be that the default forge API url set by forge-ruby would update from https://forgeapi.puppetlabs.com to https://forgeapi.puppet.com.

Comment (caseywilliams) [reply doc/dynamic-environments/configuration.mkd:171]: I guess a third option would be to add new logic in r10k to prepend `Bearer ` to the key as needed - it wouldn't affect forge-ruby behavior in 3.1.0, and it would let people omit `Bearer ` in their r10k.yaml file if they're using an older forge-ruby version.

Comment (Magisus) [reply doc/dynamic-environments/configuration.mkd:171]: We talked about that and decided we didn't know enough about what token requirements other custom Forge implementations might have, and didn't want to introduce magic that was going to bite us. Though I guess if the Forge gem is doing that now, that point is kinda moot.

Comment (mwaggett) [reply doc/dynamic-environments/configuration.mkd:171]: @caseywilliams Would the URL update in 3.1.0 that you mentioned cause breakages? I know we had some trouble updating that URL in r10k, so we're still using the `puppetlabs` one.

Comment (caseywilliams) [reply doc/dynamic-environments/configuration.mkd:171]: Yeah, it might - if I remember correctly, the PE acceptance tests were breaking last time we tested using the updated URL in r10k. forgeapi.puppet.com is really the current canonical place to find the API, so everyone should switch over eventually, but the two biggest issues in updating would be (as far as I can tell) (1) if a user had allowlisted access in their network to forgeapi.puppetlabs.com and then suddenly lost access when the url updated but their rules had not, and (2) needing to schedule the acceptance test update work

**Comment by binford2k** [comment]
@mwaggett manually tested, works as expected 👍

File: doc/dynamic-environments/configuration.mkd
Comment (mwaggett) [reply doc/dynamic-environments/configuration.mkd:171]: looks like bumping the puppet_forge gem to 3.1.0 will involve packaging a bunch of additional dependencies in pe-r10k-vanagon, since faraday 1.3+ has a bunch more dependencies than the version we're currently on.

I'm thinking it might be best to open a separate ticket/issue for bumping puppet_forge, removing the docs around `Bearer `, and updating the Forge API url. Would you mind doing that @caseywilliams?

**Review by mwaggett — APPROVED** [review]

File: doc/dynamic-environments/configuration.mkd
Comment (caseywilliams) [reply doc/dynamic-environments/configuration.mkd:171]: Sure, wasn't sure exactly which jira project was appropriate but I filed [PE-32770](https://tickets.puppetlabs.com/browse/PE-32770).

**Pull request merged**
